draft-yasskin-dispatch-bundled-exchanges.md

```diff
@@ -0,0 +1,675 @@
+1. If `bytes` includes extra bytes after the encoding of a CBOR item, return an
+   error.
+1. Let `item` be the result of decoding `bytes` as a CBOR item.
+1. If a CDDL rule was specified, but `item` does not match it, return an error.
+1. Return `item`.
+
+### Parsing variable-length data from a bytestring {#parse-bytestring}
+
+Bundles encode variable-length data in CBOR bytestrings, which are prefixed with
+their length. This algorithm returns the number of bytes in the variable-length
+item and sets the stream's current offset to  the first byte of the contents.
+
+To get the length of a CBOR bytestring header from a bundle's stream, the parser MUST do the following:
+
+1. Wait for 9 bytes to be available on the stream. If this fails, return an
+   error. Note that because of the 8-byte length at the end of a bundle, this
+   will never fail for a well-formed bundle.
+1. Let `firstByte` be the result of reading 1 byte from the stream.
+1. If `firstByte & 0xE0` is not `0x40`, the item is not a bytestring. Return an
+   error.
+1. If `firstByte & 0x1F` is:
+
+   0..23, inclusive
+   : Return `firstByte`.
+
+   24
+   : Let `content` be the result of reading 1 byte from the stream. If `content`
+     is less than 24, return an error.
+
+   25
+   : Let `content` be the result of reading 2 bytes from the stream. If the
+     first byte of `content` is 0, return an error.
+
+   26
+   : Let `content` be the result of reading 4 bytes from the stream. If the
+     first two bytes of `content` are 0, return an error.
+
+   27
+   : Let `content` be the result of reading 8 bytes from the stream. If the
+     first four bytes of `content` are 0, return an error.
+
+   28..31, inclusive
+   : Return an error.
+1. Return the big-endian integer encoded in `content`.
+
+## Interpreting CBOR HTTP headers {#cbor-headers}
+
+Bundles represent HTTP requests and responses as a list of headers, matching the
+following CDDL ({{!I-D.ietf-cbor-cddl}}):
+
+~~~ cddl
+headers = {* bstr => bstr}
+~~~
+
+Pseudo-headers starting with a `:` provide the non-header information needed to
+create a request or response as appropriate
+
+To convert a CBOR item `item` into a {{FETCH}} header list and pseudoheaders,
+parsers MUST do the following:
+
+1. If `item` doesn't match the `headers` rule in the above CDDL, return an
+   error.
+1. Let `headers` be a new header list ({{FETCH}}).
+1. Let `pseudos` be an empty map ({{INFRA}}).
+
+1. For each pair `name`/`value` in `item`:
+   1. If `name` contains any non-ASCII or upper-case characters, return an
+      error. This matches the requirement in Section 8.1.2 of {{?RFC7540}}.
+   1. If `name` starts with a ':':
+      1. If `pseudos[name]` exists, return an error.
+      1. Set `pseudos[name]` to `value`.
+      1. Continue.
+   1. If `name` or `value` doesn't satisfy the requirements for a header in
+      {{FETCH}}, return an error.
+   1. If `headers` contains ({{FETCH}}) `name`, return an error.
```

Comment:
You cannot have multiple headers of the same name? Might be worth calling out you cannot encode cookies.